src/Pos/Util/UserSecret.hs

```diff
@@ -81,19 +96,50 @@ instance Bi UserSecret where
         keys <- get
         return $ def & usVss .~ vss & usPrimKey .~ pkey & usKeys .~ keys
 
+#ifdef POSIX
+-- | Constant that defines file mode 600.
+mode600 :: PSX.FileMode
+mode600 = PSX.unionFileModes PSX.ownerReadMode PSX.ownerWriteMode
+
+-- | Check whether a given file has mode 600.
+failIfModeNot600 :: (MonadIO m, MonadFail m) => FilePath -> m ()
+failIfModeNot600 path = do
+    mode <- liftIO $ PSX.fileMode <$> PSX.getFileStatus path
+    when (mode /= mode600) $
+        fail "Secret file mode incorrect. Set it to 600 and try again."
+
+-- | Set mode 600 on a given file, regardless of its current mode.
+setMode600 :: (MonadIO m) => FilePath -> m ()
+setMode600 path = liftIO $ PSX.setFileMode path mode600
+#endif
+
 -- | Create user secret file at the given path, but only when one doesn't
 -- already exist.
 initializeUserSecret :: (MonadIO m) => FilePath -> m ()
 initializeUserSecret path = do
     exists <- T.testfile (fromString path)
-    liftIO (if exists
-            then return ()
-            else T.output (fromString path) empty)
+    liftIO $
+#ifdef POSIX
+        if exists
+        then failIfModeNot600 path
```

Comment:
this is quite racy, for something that could be done as one operation; you can create the file (O_CREAT) and fail if it already exist (O_EXCL), and the mode are automatically set to be the right one with the 3 parameters of `open`.

It's probably possible with `unix` already, but using the `handle` branch of `foundation`, you can do that with:

```haskell
import Foundation.System.Bindings.Posix

withFilepath filepath $ \fp -> do
fd <- sysPosixOpen fp (sysPosix_O_WRONLY .|. sysPosix_O_CREAT .|. sysPosix_O_EXCL .|. sysPosix_O_NONBLOCK) (CMode 0o600)
...
```